test/perf/fifo_test_multi_gpu.cu

```diff
@@ -0,0 +1,639 @@
+    std::cout << ss.str() << std::endl;
+  }
+
+  nlohmann::ordered_json combinedMetrics;
+
+  // Prepare variables for the test variant
+  mscclpp::SemaphoreId semaphoreId = 0;
+  std::shared_ptr<mscclpp::Connection> connection = nullptr;
+  mscclpp::RegisteredMemory remoteFlagRegMem = localFlagRegmem;
+  mscclpp::PortChannelDeviceHandle portChannelHandle;
+
+  if (worldSize >= 2 && !connections.empty()) {
+    int peerRank = (rank == 0) ? 1 : 0;
+    int connIndex = peerRank < rank ? peerRank : peerRank - 1;
+    if (connIndex < connections.size()) {
+      connection = connections[connIndex];
+      semaphoreId = proxyService->buildAndAddSemaphore(*communicator, connection);
+      auto portChannel = proxyService->portChannel(semaphoreId, proxyService->addMemory(localFlagRegmem),
+                                                   proxyService->addMemory(localFlagRegmem));
+      portChannelHandle = portChannel.deviceHandle();
+      cudaMemcpyToSymbol(gPortChannel, &portChannelHandle, sizeof(portChannelHandle), 0, cudaMemcpyHostToDevice);
+    }
+  }
+
+  for (int numParallel : config.parallelismLevels) {
+    // Add synchronization before each test iteration
+    MPI_Barrier(MPI_COMM_WORLD);
+
+    runFifoTestVariant(hostFifo, stream, numParallel, combinedMetrics, rank, portChannelHandle);
+
+    // Add synchronization after each test iteration
+    MPI_Barrier(MPI_COMM_WORLD);
+  }
+
+  std::map<std::string, std::string> testParams;
+  testParams["fifo_size"] = std::to_string(static_cast<int>(hostFifo->size()));
+
+  // Add parallelism levels to test parameters
+  std::stringstream parallelismStream;
+  for (size_t i = 0; i < config.parallelismLevels.size(); ++i) {
+    if (i > 0) parallelismStream << ",";
+    parallelismStream << config.parallelismLevels[i];
+  }
+  testParams["parallelism_levels"] = parallelismStream.str();
+
+  utils::recordResult(testName, "fifo", combinedMetrics, testParams);
+
+  // Cleanup
+  utils::CUDA_CHECK(cudaStreamDestroy(stream));
+  cudaFree(localSemaphoreFlag);
+
+  proxyService->stopProxy();
+}
+
+void runAllFifoTests(const mscclpp::test::TestContext& context) {
+  // clang-format off
+  std::vector<FifoTestConfig> configs = {
+      {1, {1}},
+      {128, {1, 8, 64, 128}},
+      {512, {1, 8, 64, 256, 512}},
+  };
+  // clang-format on
+
+  for (const auto& config : configs) {
+    runFifoTest(config, context);
+  }
+}
+
+// Main multi-GPU test function
+void runMultiGpuTest(const MultiGpuTestConfig& config, const mscclpp::test::TestContext& context) {
+  int rank = context.rank;
+  int worldSize = context.size;
+  auto communicator = context.communicator;
+  auto bootstrap = context.bootstrap;
+
+  if (worldSize != config.numGpus) {
+    throw std::invalid_argument("World size must match number of GPUs in config");
+  }
+
+  // Set the device for this process
+  cudaSetDevice(rank);
+
+  // Setup transport
+  mscclpp::TransportFlags transport = mscclpp::Transport::CudaIpc | mscclpp::Transport::IB0;
```

Comment:
Different devices need to use different IB interface. For example GPU-0 use IB-0, GPU 1 use IB-1